### DISCUSSION TITLE
Strings vs. numbers; writing and reading

### DISCUSSION THREAD

**Discussion by staleyLANL** [nlohmann/json#3460] [Ideas]
This may be a feature request. Or possibly it's doable already, but if so, I'm not sure how. My apologies if this has an answer somewhere already. I hope also that I haven't overlooked something that makes all of this nonsensical. I think, at least, that it makes sense.

So, we all know it's super easy to write a string value into a key/value pair:

```
nlohmann::json j;
j["foo"] = "bar";
std::cout << std::setw(2) << j << std::endl;
```

which of course prints:

```
{
  "foo": "bar"
}
```

Or, we could give a "number" (integral or floating-point):

```
nlohmann::json j;
j["foo"] = 0.000012345;
std::cout << std::setw(2) << j << std::endl;
```

which for me prints:

```
{
  "foo": 1.2345e-05
}
```

Here's the issue. I have a code that gives users very fine control of exactly how numbers are formatted, with regards to number of digits, fixed versus scientific notation, etc. Someone could utilize existing C++ standard-library constructs like `std::setprecision`, `std::fixed` and `std::scientific`, *or* could use something totally custom.

Omitting details that aren't important here, this is handled internally by converting between numbers (floating-point being the most important, but integral types are allowed as well), and *strings* into which numbers are printed *exactly* as they're wanted. From there, I support I/O from/to various text file formats - XML, a couple of custom formats, and -- I hope -- JSON.

So, I want to do something like this:

```
nlohmann::json j;
j["foo"] = "0.000012345"; // <== but not quite this right-hand-side!
std::cout << std::setw(2) << j << std::endl;
```

and get exactly this:

```
{
  "foo": 0.000012345
}
```

Two points here. (1) That's a JSON *number*, not a *quoted* string (as we'd get with the code snippet, as-is, with its string right-hand-side). NO quotes! (2) The string's contents are something that I can promise (under penalty of undefined behavior, I suppose!) is a valid *number*, and I want it printed exactly as-is, character for character. So, in this example, I want `0.000012345` in the output. Not `"0.000012345"`, and not `1.2345e-05`.

Essentially, then: "here's a string, I promise it's a valid number, print it exactly as-is without quotes".

I need this for arrays, too. I'll have an array of `std::string`, for example:

```
std::vector<std::string>{"0.00001","-0.00002"};
```

and wish to say "the strings are actually numbers" and get:

```
{
  "foo": [
    0.00001,
    -0.00002
  ]
}
```

Again, with the strings printed as numbers -- so, not quoted -- exactly as-is, and with no conversion to `double` happening under the hood. I'll just promise that each string's contents looks like a valid number.

The above is all about writing. However....

Just as importantly, I'll also need to *read* JSON files that are produced as described above. Now, if (as I've promised) the strings-printed-as-numbers are valid numbers, then a third party code should be able to read the JSON file in the customary manner, and get something reasonable. My own code, however, supports this business of allowing for the exact preservation of a number's representation, in terms of significant digits, format and such. Right now, this:

```
const nlohmann::json j = R"({"foo": 0.000012345})"_json;
const double scalar = j["foo"];
std::cout << scalar << std::endl;
```

prints `1.2345e-05`, and this:

```
const nlohmann::json j = R"({"foo": [0.00001,-0.00002]})"_json;
const std::vector<double> vector = j["foo"];
std::cout << vector[0] << std::endl;
std::cout << vector[1] << std::endl;
```

prints `1e-05` and `-2e-05`. For my own code's reading of the literals above (or of a file with such content), I'd want to be able to recover the exact text of the numbers as they appear in the input: `"0.000012345"`, `"0.00001"`, and `"-0.00002"`.

This would require, of course, that an initialization like:

```
const nlohmann::json j = R"(...)"_json;
```

(or an equivalent read-from-file) *not* automatically place such numbers into floating-points and lose the original characters. (Placing into floating-point is fine by itself, of course, as long as I can ignore those and retrieve the original characters.)

Basically, then, for reading, I want to create an `nlohmann::json` from some source (literal string, say, or file). Then, for each key/value pair, I'd wish to ask what JSON type it has. (The usual: number, string, boolean, array, object, or null). If number or array of numbers, then I'll want the retrieve the numbers as strings, directly as they appear textually in the input.

**Comment by falbrechtskirchinger** [comment]
The most obvious solution (in my mind, anyway) would be to use class-types for `number_float_t`, `number_integer_t`, and `number_unsigned_t`. Currently, that is of course not possible, as all three types are required to be primitive. But if someone wants to pursue a "proper" solution, that is my suggestion.

The part about reading numbers and retaining them as strings is doable (see this previous discussion #3432).

And if you're willing to get your hands dirty and are willing to modify the library (and maintain those modifications yourself), it shouldn't be too difficult to modify the serializer to check if a string is parsable as a number and then serialize without quotation marks. If you want some pointers, I can take a look at the code for you later. It probably requires fewer lines of code than the reading part.

Edit:
It really is pretty easy. With 4 edits each in `class serializer` and `class lexer` I get the following result:
```cpp
json j;
j["n"] = 0.0000012345;
j["n_string"] = "0.0000012345";
j["string"] = "regular string";
std::cout << j.dump(4) << "\n";
```
```json
{
    "n": 1.2345e-06,
    "n_string": 0.0000012345,
    "string": "regular string"
}
```

I'm reusing the `lexer` to determine if a string is a number. `lexer` is not designed to work on multiple inputs and also isn't designed to just lex numbers, so I had to add a `reset_input()` function, un-private `scan_number()` and replace an assertion.
In the `serializer` I've added `lexer` as a member to avoid frequent reinstantiations.

Here's the diff to get you started:
```diff
diff --git a/include/nlohmann/detail/input/lexer.hpp b/include/nlohmann/detail/input/lexer.hpp
index d0c063a1c..3b30877a0 100644
--- a/include/nlohmann/detail/input/lexer.hpp
+++ b/include/nlohmann/detail/input/lexer.hpp
@@ -125,6 +125,16 @@ class lexer : public lexer_base<BasicJsonType>
     lexer& operator=(lexer&&) = default; // NOLINT(hicpp-noexcept-move,performance-noexcept-move-constructor)
     ~lexer() = default;
 
+    void reset_input(InputAdapterType&& adapter, bool do_get = false)
+    {
+        ia = std::move(adapter);
+        position = position_t{};
+        if (do_get)
+        {
+            get();
+        }
+    }
+
   private:
     /////////////////////
     // locales
@@ -958,6 +968,7 @@ class lexer : public lexer_base<BasicJsonType>
           locale's decimal point is used instead of `.` to work with the
           locale-dependent converters.
     */
+  public:
     token_type scan_number()  // lgtm [cpp/use-of-goto]
     {
         // reset token_buffer to store the number's bytes
@@ -998,7 +1009,8 @@ class lexer : public lexer_base<BasicJsonType>
 
             // all other characters are rejected outside scan_number()
             default:            // LCOV_EXCL_LINE
-                JSON_ASSERT(false); // NOLINT(cert-dcl03-c,hicpp-static-assert,misc-static-assert) LCOV_EXCL_LINE
+                // JSON_ASSERT(false); // NOLINT(cert-dcl03-c,hicpp-static-assert,misc-static-assert) LCOV_EXCL_LINE
+                return token_type::parse_error;
         }
 
 scan_number_minus:
@@ -1283,6 +1295,8 @@ scan_number_done:
         return token_type::value_float;
     }
 
+  private:
+
     /*!
     @param[in] literal_text  the literal text to expect
     @param[in] length        the length of the passed literal text
diff --git a/include/nlohmann/detail/output/serializer.hpp b/include/nlohmann/detail/output/serializer.hpp
index b6349ea8f..3e6f65535 100644
--- a/include/nlohmann/detail/output/serializer.hpp
+++ b/include/nlohmann/detail/output/serializer.hpp
@@ -49,6 +49,9 @@ class serializer
     static constexpr std::uint8_t UTF8_ACCEPT = 0;
     static constexpr std::uint8_t UTF8_REJECT = 1;
 
+    using lexer_ia_t = iterator_input_adapter<std::string::const_iterator>;
+    using lexer_t = lexer<BasicJsonType, lexer_ia_t>;
+
   public:
     /*!
     @param[in] s  output stream to serialize to
@@ -64,6 +67,7 @@ class serializer
         , indent_char(ichar)
         , indent_string(512, indent_char)
         , error_handler(error_handler_)
+        , lex(lexer_ia_t({}, {}))
     {}
 
     // delete because of pointer members
@@ -237,8 +241,15 @@ class serializer
 
             case value_t::string:
             {
+                const auto& str = *val.m_value.string;
+                lex.reset_input(lexer_ia_t(str.begin(), str.end()), true);
+                if (lex.scan_number() != lexer_t::token_type::parse_error)
+                {
+                    dump_escaped(str, ensure_ascii);
+                    return;
+                }
                 o->write_character('\"');
-                dump_escaped(*val.m_value.string, ensure_ascii);
+                dump_escaped(str, ensure_ascii);
                 o->write_character('\"');
                 return;
             }
@@ -973,6 +984,8 @@ class serializer
 
     /// error_handler how to react on decoding errors
     const error_handler_t error_handler;
+
+    lexer_t lex;
 };
 }  // namespace detail
 }  // namespace nlohmann
```

**Reply by staleyLANL** [reply]
Ah yes, thanks for pointing me to the existing answer about *reading* numbers in terms of their original string form. I'll look into it.

For writing, I'm thinking in terms of, as you suggested, a "proper" solution that perhaps looks something like the following.

Consider that `nlohmann/json` currently allows us to write:

```
nlohmann::json j;
T value; // <== and initialize somehow
j["key"] = value;
```

for certain specific types T -- `int`, for example, and `double`, and `std::string`, and `std::vector`s thereof.

Imagine that `nlohmann/json` shunted all of its output through, say, a free function, which we'll call `toJSONValue`. Instances of `toJSONValue` for the existing supported types `T` would be in the code base, doing for `T` exactly what's being done right now.

Then, if someone has their own custom way in which they'd really like to write something, they'd just provide a suitable instance of `toJSONValue`. In my case, for dealing with "strings" that might actually represent numbers, I'd cook up something like this:

```
// Make a wrapper class, Stringified
template<class T>
struct Stringified {
   std::string rep; // <== ALWAYS a std::string, regardless of T
   explicit Stringified(const std::string &r) : rep(r) { }
};

// Next, tell nlohmann/json exactly how to write a Stringified...

// General case: assume a number
template<class T>
inline std::string toJSONValue(const Stringified<T> &obj)
{
   return obj.rep; // no quotes!
}

// Special case: is really a std::string
inline std::string toJSONValue(const Stringified<std::string> &obj)
{
   return '"' + obj.rep + '"'; // quotes just for std::string!
}
```

Various signatures could be considered for the `toJSONValue` functions, depending on what would work most smoothly in `nlohmann/json`. Above, they return the value's desired representation as a `std::string`, which I suppose `nlohmann/json` would probably ship on to some `std::ostream` into which it's actually writing content.

Another design might be for `toJSONValue` to receive both the value, and a `std::ostream` into which to write it directly. If, in any context, `nlohmann/json` isn't dealing with an existing `std::ostream` and does need a `std::string`, it can of course just send `toJSONValue` a `std::ostringstream` intermediary, then get its `.str()`.

One could also imagine `nlohmann/json` having a *default* `toJSONValue` that forwards to a C++ stream output operator for the object, so that instead of specializing `toJSONValue` for my `Stringified` class, I could write:

```
// For general T
template<class T>
inline std::ostream &operator<<(std::ostream &os, const Stringified<T> &obj)
{
   return os << obj.rep;
}

// For std::string
inline std::ostream &operator<<(std::ostream &os, const Stringified<std::string> &obj)
{
   return os << '"' << obj.rep << '"';
}
```

and those would just work, right out of the box.

Finishing my thought above, and using a "number" from my original post, someone could now write:

```
nlohmann::json j;
j["aDouble"] = Stringified<double>{"0.000012345"}; // intended as a number
j["aString"] = Stringified<std::string>{"abcdef"}; // an actual string
std::cout << std::setw(2) << j << std::endl;
```

and get:

```
{
  "aDouble": 0.000012345,
  "aString": "abcdef"
}
```

which is exactly what we're looking for. (Of course, I suppose there's no real need for the `std::string` specialization of Stringified, as one could always use a `std::string` directly, but we'd have it for completeness.)

For vectors, nothing special would be required. The `toJSONValue` business would be applied to each element.

A solution as described above would, of course, be a "know what you're doing!" sort of deal, as `nlohmann/json` can't possibly know if somebody is going to produce a reasonable, valid JSON value.

I'll study the other discussion, regarding *reading* rather than *writing*. Offhand, for reading, I might imagine having something like `fromJSONValue` as the *input* analog to my `toJSONValue` -- to simplify the process, and to make specialized input and output function in a symmetric manner.

I probably wouldn't want to modify the serializer so that it tries to guess whether a string is parsable as a number. An external code would presumably do that already - my own code does - and it wouldn't really be the job of `nlohmann/json`. For that matter, we can probably all agree that `"<looks like a number>"` is properly regarded as a string, even if it looks like a number. We'd want to use the C++ type system somehow, as with our "proper" solution, to force special handling. That way, existing handling of strings vs. numbers isn't touched, as it shouldn't be.

**Reply by falbrechtskirchinger** [reply]
```
j["aDouble"] = Stringified<double>{"0.000012345"}; // intended as a number
```

Unfortunately, this won't work as you envision. This code will in fact go through the existing `to_json` machinery that will try to convert your type to one of the supported types (i.e., `std::vector`, `std::string`, `double`, etc.) for storage inside the `json` object. If no `to_json` function for `Stringified<double>` exists and `Stringified<double>` isn't convertible to a supported type, it will simply fail to compile.

What I suggest is to allow arbitrary class types for numbers as part of the `basic_json` template.
That would enable you to implement a number class that is convertible to `double`, and therefore behaves like a double where needed, but can also store a string representation for serialization.

I don't know how feasible this option ultimately is either, but it is where I'd start.
